synedrion/src/cggmp21/sigma/mod_.rs

```diff
@@ -135,34 +136,37 @@ impl<P: SchemeParams> ModProof<P> {
         }
     }
 
-    pub fn verify(
-        &self,
-        rng: &mut impl CryptoRngCore,
-        pk: &PublicKeyPaillier<P::Paillier>,
-        aux: &impl Hashable,
-    ) -> bool {
+    pub fn verify(&self, pk: &PublicKeyPaillier<P::Paillier>, aux: &impl Hashable) -> bool {
         let challenge = ModChallenge::new(pk, &self.commitment, aux);
         if challenge != self.challenge {
             return false;
         }
 
+        let mut reader = XofHasher::new_with_dst(b"P_mod RNG").chain(aux).finalize_to_reader();
+        let mut seed = <ChaCha8Rng as SeedableRng>::Seed::default();
+        reader.read(&mut seed);
+        let mut rng = ChaCha8Rng::from_seed(seed);
+
         // The paper requires checking that `N` is odd here,
         // but it is already an invariant of `PublicKeyPaillier`.
+        if (*pk.modulus()).is_prime_with_rng(&mut rng) {
```

Comment:
I guess I don't understand why go through the primality check rather than checking the LSB and run the GCD check. Can you elaborate?

Reply:
The condition in the paper says `N is an odd composite number`, so I think we have to check it.

Reply:
Right, the modulus is untrusted input so we don't know it's `p*q`... of course.